Add salmon

sainirmayi opened the pull request:
## Description
Add salmon index and quant components

<!-- Provide a list of changes made in this PR -->

## Issue ticket number

Closes #xxxx 

<!-- Replace xxxx with the GitHub issue number -->

## Checklist before requesting a review

- [x] I have performed a self-review of my code

- [x] Conforms to the [Contributing guidelines](https://github.com/viash-hub/base/blob/main/CONTRIBUTING.md)

- [x] Proposed changes are described in the [CHANGELOG.md](https://github.com/viash-hub/base/blob/main/CHANGELOG.md)

- [x] I have tested my code with `viash ns test --parallel -q <name or namespace>`

- Check the correct box. Does this PR contain:
  - [ ] Breaking changes
  - [x] New functionality
  - [ ] Major changes
  - [ ] Minor changes
  - [ ] Documentation
  - [ ] Bug fixes

<!-- Thank you for your contribution! -->

rcannood commented:
I merged the PR that upgrades the viash version in biobase to 0.9. Could you update this PR to the Viash 0.9 format? :bow: Thanks!

rcannood commented on src/salmon/quant/test_data/salmon/transcriptome_index/complete_ref_lens.bin:
Rather than upload the whole `transcriptome_index`, would it be sufficient to just upload the reference fasta and gtf and generate the index from within the test?

Example: https://github.com/viash-hub/biobase/pull/22/files#diff-fd756488e58d46c8655c4f6a0822a25b5b5adadf7b26e9785b2d5244f3149047R75-R82

sainirmayi replied:
There weren't any other reference files available to generate the index. There is the transcriptome.fasta that I used as test data for the `salmon_index` component. But I can't be certain that it's the same one used to generate this index.

rcannood commented on src/salmon/quant/test_data/reads/a_1.fq.gz:
Can you make `reads/*.fq.gz` smaller? For example, you could use reads for which you know whether or not they will map to the reference. Example: https://github.com/viash-hub/biobase/pull/22/files#diff-fd756488e58d46c8655c4f6a0822a25b5b5adadf7b26e9785b2d5244f3149047R41-R61 . SEQ_ID1 maps perfectly to the provided genome.fasta, while SEQ_ID2 doesn't map at all.

sainirmayi replied:
I don't have the reference genome so this might not be possible

rcannood commented on src/salmon/quant/config.vsh.yaml:
```diff
@@ -0,0 +1,582 @@
+
+  - name: Output
+    arguments:
+      - name: --output
+        alternatives: ["-o"]
+        type: file
+        direction: output
+        description: |
+          Output quantification directory.
+        required: true
+        example: quant_output
```
Can you split up this output directory? For example, the star_align_reads component looks at certain files it expects and will copy them to the desired output path : https://github.com/viash-hub/biobase/pull/22/files#diff-a13e23ef40c5889bceacefe5ae19749cb879ce3a2bab40b6145900e201af6d5bR54-R62 . If we output a single directory, it makes it harder for nextflow workflows to find out which files are available or not.

It's fine to also allow outputting the whole directory, as long as it's not required to do it this way.

sainirmayi replied:
I've changed it to split the main output, i.e. the quantification file. So the output now consists of the output directory and a separate quantification file.

rcannood commented on src/salmon/index/test.sh:
```diff
@@ -0,0 +1,17 @@
+  --transcripts $dir_in/transcriptome.fasta \
+  --index index
+
+echo ">>> Checking whether output exists"
+[ ! -d "index" ] && echo "'index' does not exist!" && exit 1
+[ -z "$(ls -A 'index')" ] && echo "'index' is empty!" && exit 1
```
Could you include some checks on whether the output is correct and not just empty?

sainirmayi replied:
Added a check to verify whether the `info.json` file exists. This file should be generated if the run was successful.

rcannood commented on src/salmon/quant/test.sh:
```diff
@@ -0,0 +1,42 @@
+  --mates2 $dir_in/reads/a_2.fq.gz $dir_in/reads/b_2.fq.gz \
+  --output "quant_pe_rep_results"
+
+echo ">> Checking output"
+[ ! -d "quant_pe_rep_results" ] && echo "Output directory quant_pe_rep_results does not exist" && exit 
+[ ! -s "quant_pe_rep_results/quant.sf" ] && echo "Output file quant_pe_rep_results/quant.sf is empty" && exit 1
```
Can you include some checks on whether the output is correct and not just not empty?

sainirmayi replied:
Added checks to verify if the main quantification file has the right columns.

rcannood commented on src/salmon/quant/config.vsh.yaml:
```diff
@@ -0,0 +1,582 @@
+          List of files containing unmated reads of (e.g. single-end reads).
+        required: false
+        multiple: true
+        multiple_sep: " "
```
Could you change all multiple_sep's into `;`?

This is because other values for `multiple_sep` such as `:`, ` `, or `,` might not always work for what is otherwise a valid file path.

sainirmayi replied:
Changing it to `;` doesn't work for some reason but I haven't been able to figure out why exactly.

rcannood reviewed (changes requested):
Thanks for the PR! I added some comments.

rcannood reviewed (commented):
Hey @sainirmayi! 

How are things going with this PR?

sainirmayi commented:
> Hey @sainirmayi!
> 
> How are things going with this PR?

1. I've added additional checks to verify if the outputs are correct also split the output directory such that the main quantification file is output separately. 
2. I tried changing `multiple_sep` into `;` but keep running into errors when I do this. 
3. I wasn't able to make the read files smaller because I don't have the reference genome to check which reads might map to the genome. Also, I don't have the reference files needed to build the index while running the test for salmon quant for the test data used here.

rcannood commented on src/salmon/quant/test_data/script.sh:
```diff
@@ -0,0 +1,9 @@
+
+# Test data was obtained from https://github.com/snakemake/snakemake-wrappers/tree/master/bio/salmon/quant/test
+
+if [ ! -d /tmp/snakemake-wrappers ]; then
+  git clone --depth 1 --single-branch --branch master https://github.com/snakemake/snakemake-wrappers /tmp/snakemake-wrappers
+fi
+
+cp -r /tmp/snakemake-wrappers/bio/salmon/quant/test/* src/salmon/quant/test_data
```
Can you manually create test files for `salmon_quant` and `salmon_index` to create some smaller fastq files:

- Have much smaller file sizes
- That allows verifying whether the counts resulting from `salmon_quant` make sense?

You can use the code I wrote [here](https://github.com/viash-hub/biobase/blob/main/src/star/star_align_reads/test.sh#L41-L82) as an example.

sainirmayi replied:
Done!

rcannood commented on src/salmon/index/config.vsh.yaml:
Could you move the `src/salmon/index` folder to `src/salmon/salmon_index`?

rcannood reviewed (approved):
LGTM!
